peps/pep-0778.rst

```diff
@@ -0,0 +1,357 @@
+PEP: 778
+Title: Safe external debugger interface for CPython
+Author: Pablo Galindo Salgado <pablogsal@python.org>, Matt Wozniski <godlygeek@gmail.com>, Ivona Stojanovic <stojanovic.i@hotmail.com>
+Discussions-To: xxxxx
+Status: Draft
+Type: Standards Track
+Topic: Interpreter Core
+Created: 25-Nov-2024
+Python-Version: 3.14
+
+Abstract
+========
+
+This PEP proposes adding a zero-overhead debugging interface to CPython that
+allows debuggers and profilers to safely attach to running Python processes. The
+interface provides safe execution points for attaching debugger code without
+modifying the interpreter's normal execution path or adding runtime overhead.
+
+A key application of this interface will be enabling pdb to attach to live
+processes by process ID, similar to ``gdb -p``, allowing developers to inspect and
+debug Python applications interactively in real-time without stopping or
+restarting them.
+
+Motivation
+==========
+
+
+Debugging Python processes in production and live environments presents unique
+challenges. Developers often need to analyze application behavior without
+stopping or restarting services, which is especially crucial for
+high-availability systems. Common scenarios include diagnosing deadlocks,
+inspecting memory usage, or investigating unexpected behavior in real-time.
+
+Very few Python tools can attach to running processes, primarily because doing
+so requires deep expertise in both operating system debugging interfaces and
+CPython internals. While C/C++ debuggers like gdb and lldb can attach to
+processes using well-understood techniques, Python tools must implement all of
+these low-level mechanisms plus handle additional complexity. For example, when
+gdb needs to execute code in a target process, it:
+
+1. Uses ptrace to allocate a small chunk of executable memory (easier said than done)
+2. Writes a small sequence of machine code - typically a function prologue, the
+   desired instructions, and code to restore registers
+3. Saves all the target thread's registers
+4. Changes the instruction pointer to the injected code
+5. Lets the process run until it hits a breakpoint at the end of the injected code
+6. Restores the original registers and continues execution
+
+Python tools face this same challenge of code injection, but with an additional
+layer of complexity. Not only do they need to implement the above mechanism,
+they must also understand and safely interact with CPython's runtime state,
+including the interpreter loop, garbage collector, thread state, and reference
+counting system. This combination of low-level system manipulation and
+deep domain specific interpreter knowledge makes implementing Python debugging tools
+exceptionally difficult.
+
+The few tools (see for example [1]_ and [2]_) that do attempt this resort to
+suboptimal and unsafe methods,
```

Comment:
The footnotes are just links and no extra info, so let's save our readers a click and link directly to them:
```suggestion
The few tools (see for example `DebugPy
<https://github.com/microsoft/debugpy/blob/43f41029eabce338becbd1fa1a09727b3cfb1140/src/debugpy/_vendored/pydevd/pydevd_attach_to_process/linux_and_mac/attach.cpp#L4>`__
and `Memray
<https://github.com/bloomberg/memray/blob/main/src/memray/_memray/inject.cpp>`__)
that do attempt this resort to suboptimal and unsafe methods,
```